### PR TITLE
Add support for sample pushdown

### PR DESCRIPTION
**Pull request by Praveen2112**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/metadata/Metadata.java [hunk 1]
```diff
@@ -398,4 +398,6 @@
     Optional<LimitApplicationResult<TableHandle>> applyLimit(Session session, TableHandle table, long limit);
 
     Optional<ConstraintApplicationResult<TableHandle>> applyFilter(Session session, TableHandle table, Constraint constraint);
+
+    Optional<TableHandle> applySample(Session session, TableHandle table, double sampleRatio);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/metadata/Metadata.java]: This should handle SYSTEM and BERNOULLI sampling. Some connectors (e.g., jdbc-based ones), may be able to support the more precise bernoulli sampling.

Comment (Praveen2112) [reply presto-main/src/main/java/io/prestosql/metadata/Metadata.java]: Have updated it to pass the type of sampling along with the ratio.

File: presto-spi/src/main/java/io/prestosql/spi/connector/SampleInformation.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.connector;
+
+import static java.util.Objects.requireNonNull;
+
+public class SampleInformation
```
Comment (martint) [inline presto-spi/src/main/java/io/prestosql/spi/connector/SampleInformation.java]: I would just pass the two fields directly to the applySample method instead of wrapping it in a separate object.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushSampleIntoTableScan.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public Rule.Result apply(SampleNode sample, Captures captures, Rule.Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        SampleInformation sampleInformation = new SampleInformation(getSamplingType(sample), sample.getSampleRatio());
+        return metadata.applySample(context.getSession(), tableScan.getTable(), sampleInformation)
+                .map(result -> {
+                    PlanNode node = new TableScanNode(
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushSampleIntoTableScan.java]: Inline the `node` variable. Then, you can simplify this to:

```java
.map(result -> Result.ofPlanNode(new TableScanNode(
        tableScan.getId(),
        result,
        tableScan.getOutputSymbols(),
        tableScan.getAssignments(),
        tableScan.getEnforcedConstraint())))
```

File: presto-main/src/main/java/io/prestosql/sql/planner/PlanOptimizers.java [hunk 4]
```diff
@@ -307,19 +307,30 @@ public PlanOptimizers(
...
                                         new PruneCountAggregationOverScalar(),
                                         new PruneOrderByInAggregation(metadata.getFunctionRegistry()),
                                         new RewriteSpatialPartitioningAggregation(metadata)))
                                 .build()),
+                new IterativeOptimizer(
+                        ruleStats,
+                        statsCalculator,
+                        estimatedExchangesCostCalculator,
+                        ImmutableSet.of(
+                                new PushLimitIntoTableScan(metadata),
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/PlanOptimizers.java]: Don't move these. There are benefits to having them in the same loop as the previous rules.

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java [hunk 5]
```diff
@@ -361,7 +364,19 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
...
+                new MemoryTableHandle(table.getId(), OptionalLong.of(limit), OptionalDouble.empty()),
                 true));
     }
+
+    @Override
+    public Optional<ConnectorTableHandle> applySample(ConnectorSession session, ConnectorTableHandle handle, SampleInformation sampleInformation)
+    {
+        MemoryTableHandle table = (MemoryTableHandle) handle;
+
+        if (table.getSampleRatio().isPresent() || sampleInformation.getSampleType() != SYSTEM) {
```
Comment (martint) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java]: If sampling has been pushed down, you can combine with the previous one by multiplying it. If they are equal, then you want to bail out.

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemorySplitManager.java [hunk 6]
```diff
@@ -65,6 +68,13 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transactionHand
                 splits.add(new MemorySplit(table.getId(), i, splitsPerNode, dataFragment.getHostAddress(), rows, OptionalLong.empty()));
             }
         }
+
+        if (table.getSampleRatio().isPresent()) {
+            double sampleRatio = table.getSampleRatio().getAsDouble();
+            return new FixedSplitSource(splits.build().stream()
+                    .filter(value -> ThreadLocalRandom.current().nextDouble() > sampleRatio)
```
Comment (martint) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemorySplitManager.java]: This should be `<`

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java (hunk 5 above)
Comment (martint) [reply presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java]: Also, this is a little tricky. If limit was pushed down first, the sampling has to be over the table+limit. If table sample is pushed first, the limit has to be over table+sample. The logic in MemorySplitManager always does sample before limit.

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java [hunk 7]
```diff
@@ -361,7 +364,20 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
...
                 true));
     }
+
+    @Override
+    public Optional<ConnectorTableHandle> applySample(ConnectorSession session, ConnectorTableHandle handle, SampleType sampleType, double sampleRatio)
+    {
+        MemoryTableHandle table = (MemoryTableHandle) handle;
+
+        if ((table.getSampleRatio().isPresent() && table.getSampleRatio().getAsDouble() == sampleRatio) ||
+                sampleType != SYSTEM) {
```
Comment (martint) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryMetadata.java]: Place this in the previous line

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemorySplitManager.java [hunk 8]
```diff
@@ -65,6 +68,13 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transactionHand
                 splits.add(new MemorySplit(table.getId(), i, splitsPerNode, dataFragment.getHostAddress(), rows, OptionalLong.empty()));
             }
         }
+
+        if (table.getSampleRatio().isPresent()) {
```
Comment (martint) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemorySplitManager.java]: This still has a semantic issue. It treats the following queries as equivalent:

```sql
SELECT *
FROM t TABLESAMPLE SYSTEM (50) 
LIMIT 1
```

vs

```sql
SELECT * FROM (
    SELECT *
    FROM t 
    LIMIT 1
)
TABLESAMPLE SYSTEM (50) 
```

For a table with a large number of rows, the first query should always return one row, while the second one should return 1 or 0 rows with a ~50% chance. So, the order of application of sample vs limit matters, but the current code doesn't take that into account.

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryPagesStore.java [hunk 9]
```diff
@@ -98,13 +101,17 @@ public synchronized void add(Long tableId, Page page)
         boolean done = false;
         long totalRows = 0;
         for (int i = partNumber; i < tableData.getPages().size() && !done; i += totalParts) {
-            Page page = tableData.getPages().get(i);
+            if (sampleRatio.isPresent() && ThreadLocalRandom.current().nextDouble() >= sampleRatio.getAsDouble()) {
```
Comment (martint) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryPagesStore.java]: This needs to be `< sampleRatio`. E.g., if sample rate is 90%, only in 10% of the cases a uniform random between [0, 1] will be higher than 0.9.

Also, since we're sampling pages, we shouldn't sample splits (in MemorySplitManager). Otherwise, we'll end up with rate^2 sampling rate (50% of splits, 50% of pages per split ~ 25% of rows overall)

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryPagesStore.java [hunk 10]
```diff
@@ -98,13 +101,17 @@ public synchronized void add(Long tableId, Page page)
...
+                continue;
+            }
 
+            Page page = tableData.getPages().get(i);
             totalRows += page.getPositionCount();
             if (limit.isPresent() && totalRows > limit.getAsLong()) {
                 page = page.getRegion(0, (int) (page.getPositionCount() - (totalRows - limit.getAsLong())));
                 done = true;
             }
+
```
Comment (martint) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryPagesStore.java]: Unrelated formatting change

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemorySplitManager.java [hunk 11]
```diff
@@ -65,6 +68,7 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transactionHand
                 splits.add(new MemorySplit(table.getId(), i, splitsPerNode, dataFragment.getHostAddress(), rows, OptionalLong.empty()));
             }
         }
+
```
Comment (martint) [inline presto-memory/src/main/java/io/prestosql/plugin/memory/MemorySplitManager.java]: Unrelated formatting change

File: presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryPagesStore.java (hunk 9 above)
Comment (martint) [reply presto-memory/src/main/java/io/prestosql/plugin/memory/MemoryPagesStore.java]: Actually, nevermind. This is the condition for skipping the page (`continue`, below), so it's correct.

**Comment by martint** [comment]
Good stuff! Merged!
